pynamodb/models.py

```diff
@@ -169,40 +171,104 @@ class MetaModel(AttributeContainerMeta):
-                        setattr(attr_obj, 'request_timeout_seconds', get_settings_value('request_timeout_seconds'))
-                    if not hasattr(attr_obj, 'base_backoff_ms'):
-                        setattr(attr_obj, 'base_backoff_ms', get_settings_value('base_backoff_ms'))
-                    if not hasattr(attr_obj, 'max_retry_attempts'):
-                        setattr(attr_obj, 'max_retry_attempts', get_settings_value('max_retry_attempts'))
+                    continue
                 elif issubclass(attr_obj.__class__, (Index, )):
                     attr_obj.Meta.model = cls
+                    has_indexes = True
                     if not hasattr(attr_obj.Meta, "index_name"):
                         attr_obj.Meta.index_name = attr_name
                 elif issubclass(attr_obj.__class__, (Attribute, )):
                     if attr_obj.attr_name is None:
                         attr_obj.attr_name = attr_name
 
-            if META_CLASS_NAME not in attrs:
-                setattr(cls, META_CLASS_NAME, DefaultMeta)
+
+            exception_attrs = {'__module__': attrs.get('__module__')}
 
             # create a custom Model.DoesNotExist derived from pynamodb.exceptions.DoesNotExist,
             # so that "except Model.DoesNotExist:" would not catch other models' exceptions
-            if 'DoesNotExist' not in attrs:
-                exception_attrs = {'__module__': attrs.get('__module__')}
-                if hasattr(cls, '__qualname__'):  # On Python 3, Model.DoesNotExist
-                    exception_attrs['__qualname__'] = '{}.{}'.format(cls.__qualname__, 'DoesNotExist')
-                cls.DoesNotExist = type('DoesNotExist', (DoesNotExist, ), exception_attrs)
-
+            class DoesNotExist(pynamodb.exceptions.DoesNotExist):
+                pass
+            cls.DoesNotExist = type('DoesNotExist', (DoesNotExist,), exception_attrs)
+
+            # create a custom Model.NotAllowedWhenAbstract derived from pynamodb.exceptions.NotAllowedWhenAbstract,
+            # so that "except Model.NotAllowedWhenAbstract:" would not catch other models' exceptions
+            class NotAllowedWhenAbstract(pynamodb.exceptions.NotAllowedWhenAbstract):
+                pass
+            cls.NotAllowedWhenAbstract = type('NotAllowedWhenAbstract', (NotAllowedWhenAbstract,), exception_attrs)
+
+            if is_abstract is True:
+                if has_table_name:
+                    raise cls.NotAllowedWhenAbstract("Abstract model can't have a table name")
+                if has_indexes:
+                    raise cls.NotAllowedWhenAbstract("Abstract model can't have an index")
+            elif is_abstract is False or is_abstract is None:
+                # Enforce that non abstract tables have table_name
+                if not has_table_name and not parent_is_attribute_container:
+                    raise InheritanceError("Non abstract models need a table name")
```

Comment:
This also seems like a breaking change -- I believe we allow dynamic reconfiguration of Models today, and as part of that allow Model's to overwrite or redefine their meta settings up until the point that we need to fetch the table metadata.

That is we can create a Model class without `table_name` and set it after class creation dynamically as long as we do so before `_get_connection` is ever called.